libasn1fix/asn1fix_cws.c

```diff
@@ -229,74 +219,94 @@ _asn1f_assign_cell_value(arg_t *arg, struct asn1p_ioc_row_s *row, struct asn1p_i
 		return -1;
 	}
 
-	p_dot = memchr(buf, '.', bend - buf);
-	if (p_dot) {
-		/* parse Module.xxxx */
-		*p_dot = '\0';
-		p = malloc(bend - p_dot);
-		assert(p);
-		memcpy(p, p_dot + 1, bend - p_dot - 1);
-		p[bend - p_dot - 1] = '\0';
+	p = malloc(bend - buf + 1);
+	assert(p);
+	memcpy(p, buf, bend - buf);
+	p[bend - buf] = '\0';
+	/* remove trailing space */
+	for (i = bend - buf - 1; (i > 0) && isspace(p[i]); i--)
+		p[i] = '\0';
+
+	/* This value 100 should be larger than following formatting string */
+	psize = bend - buf + 100;
+	pp = malloc(psize);
+	if(pp == NULL) {
+		free(p);
+		return -1;
+	}
 
-		ext_module = asn1f_lookup_module(arg, (const char *)buf, 0);
+	if(cell->field->expr_type == A1TC_CLASSFIELD_TFS) {
+		ret = snprintf(pp, psize,
+			"M DEFINITIONS ::=\nBEGIN\n"
+			"V ::= %s\n"
+			"END\n",
+			p
+		);
+	} else if(cell->field->expr_type == A1TC_CLASSFIELD_FTVFS) {
+		type_expr = TQ_FIRST(&(cell->field->members));
+		ret = snprintf(pp, psize,
+				"M DEFINITIONS ::=\nBEGIN\n"
+				"v %s ::= %s\n"
+				"END\n",
+				type_expr->reference ? 
+					type_expr->reference->components[0].name : 
+					_asn1p_expr_type2string(type_expr->expr_type),
+				p
+			);
 	} else {
-		p = malloc(bend - buf + 1);
-		assert(p);
-		memcpy(p, buf, bend - buf);
-		p[bend - buf] = '\0';
+		free(p);
+		free(pp);
+		WARNING("asn1c only be able to parse TypeFieldSpec and FixedTypeValueFieldSpec. Failed when parsing %s at line %d\n", p, arg->expr->_lineno);
+		return -1;
 	}
-
-	if(!isalpha(*p)) {
-		if(isdigit(*p)) {
-			asn1c_integer_t value;
-			if(asn1p_atoi(p, &value)) {
-				FATAL("Value %s at line %d is too large for this compiler! Contact the asn1c author.\n", p, arg->expr->_lineno);
-				free(p);
-				return -1;
-			}
-			expr = asn1p_expr_new(arg->expr->_lineno, arg->expr->module);
-			expr->Identifier = p;
-			expr->meta_type = AMT_VALUE; 
-			expr->expr_type = ASN_BASIC_INTEGER;
-			expr->value = asn1p_value_fromint(value);
-		} else if (*p == '"') {
-			expr = asn1p_expr_new(arg->expr->_lineno, arg->expr->module);
-			expr->Identifier = p;
-			expr->meta_type = AMT_VALUE;
-			expr->expr_type = ASN_BASIC_CHARACTER_STRING;
-			expr->value = asn1p_value_frombuf(p + 1, strlen(p) - 2, 0);
-		} else {
-			WARNING("asn1c is not yet able to parse arbitrary direct values; try converting %s at line %d to a reference.", p, arg->expr->_lineno);
-			free(p);
-			return 1;
-		}
+	DEBUG("ASN.1 :\n\n%s\n", pp);
+
+	assert(ret < psize);
+	psize = ret;
+
+	asn = asn1p_parse_buffer(pp, psize, A1P_NOFLAGS);
+	free(pp);
+	if(asn == NULL) {
+		FATAL("Cannot parse Setting token %s "
+			"at line %d",
+			p,
+			arg->expr->_lineno
+		);
+		free(p);
+		return -1;
 	} else {
-		ref = asn1p_ref_new(arg->expr->_lineno, arg->expr->module);
-		asn1p_ref_add_component(ref, p, RLT_UNKNOWN);
-		assert(ref);
```

Comment:
And here `ref` variable got its assignment - again, the code your #23 tried to correct.